Framework/CurveFitting/src/Algorithms/DoublePulseFit.cpp

```diff
@@ -0,0 +1,405 @@
+
+  convolution->setAttributeValue("FixResolution", false);
+  delta1->setParameter("Centre", -offset);
+  delta1->setParameter("Height", firstPulseWeight);
+  delta1->fixAll();
+  delta2->setParameter("Centre", 0.0);
+  delta2->setParameter("Height", secondPulseWeight);
+  delta2->fixAll();
+  deltaComposite->addFunction(delta1);
+  deltaComposite->addFunction(delta2);
+
+  convolution->addFunction(clonedFunction);
+  convolution->addFunction(deltaComposite);
+
+  return convolution;
+}
+
+Mantid::API::IFunction_sptr getDoublePulseFunction(
+    std::shared_ptr<const API::MultiDomainFunction> const &initialFunction,
+    double offset, double firstPulseWeight, double secondPulseWeight) {
+  auto doublePulseFunc = std::make_shared<API::MultiDomainFunction>();
+  for (size_t domain = 0; domain < initialFunction->getNumberDomains();
+       domain++) {
+    auto innerFunction = std::dynamic_pointer_cast<API::ParamFunction>(
+        initialFunction->getFunction(domain));
+    if (innerFunction) {
+      auto twoPulseInnerFunction = getDoublePulseFunction(
+          innerFunction, offset, firstPulseWeight, secondPulseWeight);
+      doublePulseFunc->addFunction(twoPulseInnerFunction);
+      doublePulseFunc->setDomainIndex(domain, domain);
+    } else {
+      throw std::runtime_error("Convolution does not support composite "
+                               "functions. DoublePulseFit.cpp");
+    }
+  }
+  return doublePulseFunc;
+}
+
+Mantid::API::IFunction_sptr extractInnerFunction(
+    std::shared_ptr<const Mantid::CurveFitting::Functions::Convolution> const
+        &convolutionFunction) {
+  return convolutionFunction->getFunction(0);
+}
+
+Mantid::API::IFunction_sptr extractInnerFunction(
+    std::shared_ptr<const API::MultiDomainFunction> const &doublePulseFunc) {
+  auto extractedFunction = std::make_shared<API::MultiDomainFunction>();
+  for (size_t domain = 0; domain < doublePulseFunc->getNumberDomains();
+       domain++) {
+    auto convFunction = std::dynamic_pointer_cast<Convolution>(
+        doublePulseFunc->getFunction(domain));
+    if (convFunction) {
+      extractedFunction->addFunction(extractInnerFunction(convFunction));
+      extractedFunction->setDomainIndex(domain, domain);
+    } else {
+      throw std::runtime_error(
+          "Cannot extract from non convolution function. DoublePulseFit.cpp");
+    }
+  }
+  return extractedFunction;
+}
+
+// Register the class into the algorithm factory
+DECLARE_ALGORITHM(DoublePulseFit)
+
+/// Default constructor
+DoublePulseFit::DoublePulseFit() : IFittingAlgorithm() {}
+
+/** Initialisation method
+ */
+void DoublePulseFit::initConcrete() {
+  declareProperty("Ties", "", Kernel::Direction::Input);
+  getPointerToProperty("Ties")
+      ->setDocumentation("Math expressions defining ties between parameters of "
+                         "the fitting function.");
+  declareProperty("Constraints", "", Kernel::Direction::Input);
```

Comment:
I believe ties and constraints can be passed as part of the fitting function. Is that right @alicerussell1 ?

Reply:
I think they can be yes. In the Fit algorithm they can also be passed in as separate properties however. I am attempting to mimic the interface of the Fit algorithm here so the two can be sued interchangeably. So I have included all the properties it has here to pass on.